### PR TITLE
[search] refactor the search logic

### LINKED ISSUE #835: [Search] Results are not limited to specified directory
When including a search component through an Experience Fragment, the results are not limited to the directory specified in the configuration.

### LINKED ISSUE #1249: Unclosed ResourceResolver in SearchResultServlet
## Bug Report

**Current Behavior**
When a search is performed using the SearchResultServlet a `ResourceResolver` is left open.
This is a known issue when using the `QueryBuilder` service. 

**Expected behavior/code**
There should be no unclosed `ResourceResolver` and no stacktrace should be shown in the logs.

**Environment**
- AEM version and patch level: 6.5.6.0
- Core Components version: 2.12.1
- JRE version: 11

**Possible Solution**
A best-attempt should be made to close the `ResourceResolver` once the query is complete.

**Additional context / Screenshots**
Log:

> 25020-07.11.2020 16:35:48.765 *INFO* [Apache Sling Resource Resolver Finalizer Thread] org.apache.sling.resourceresolver.impl.CommonResourceResolverFactoryImpl Unclosed ResourceResolver was created here: 
25021-java.lang.Exception: Opening Stacktrace
25022-	at org.apache.sling.resourceresolver.impl.CommonResourceResolverFactoryImpl$ResolverReference.<init>(CommonResourceResolverFactoryImpl.java:540) [org.apache.sling.resourceresolver:1.6.8]
25023-	at org.apache.sling.resourceresolver.impl.CommonResourceResolverFactoryImpl.register(CommonResourceResolverFactoryImpl.java:235) [org.apache.sling.resourceresolver:1.6.8]
25024-	at org.apache.sling.resourceresolver.impl.ResourceResolverImpl.<init>(ResourceResolverImpl.java:104) [org.apache.sling.resourceresolver:1.6.8]
25025-	at org.apache.sling.resourceresolver.impl.ResourceResolverImpl.<init>(ResourceResolverImpl.java:97) [org.apache.sling.resourceresolver:1.6.8]
25026-	at org.apache.sling.resourceresolver.impl.CommonResourceResolverFactoryImpl.getResourceResolverInternal(CommonResourceResolverFactoryImpl.java:280) [org.apache.sling.resourceresolver:1.6.8]
25027-	at org.apache.sling.resourceresolver.impl.CommonResourceResolverFactoryImpl.getResourceResolver(CommonResourceResolverFactoryImpl.java:190) [org.apache.sling.resourceresolver:1.6.8]
25028-	at org.apache.sling.resourceresolver.impl.ResourceResolverFactoryImpl.getResourceResolver(ResourceResolverFactoryImpl.java:98) [org.apache.sling.resourceresolver:1.6.8]
25029-	at com.day.cq.search.impl.builder.QueryBuilderImpl.createResourceResolver(QueryBuilderImpl.java:210) [com.day.cq.cq-search:5.12.16]
25030-	at com.day.cq.search.impl.builder.QueryImpl.getResourceResolver(QueryImpl.java:240) [com.day.cq.cq-search:5.12.16]
25031-	at com.day.cq.search.impl.result.HitImpl.getResource(HitImpl.java:108) [com.day.cq.cq-search:5.12.16]
25032-	at com.adobe.cq.wcm.core.components.internal.servlets.SearchResultServlet.getResults(SearchResultServlet.java:203)
25033-	at com.adobe.cq.wcm.core.components.internal.servlets.SearchResultServlet.doGet(SearchResultServlet.java:101)

### PR DESCRIPTION
**Pull request by jckautzmann**
Adds to PR #1253
Fixes #1249

Also adds to PR #1348
which fixes #835

### CODE REVIEW

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/SearchResultServlet.java [hunk 1]
```diff
@@ -16,282 +16,352 @@
...
+     * Gets the search component for the given request.
+     *
+     * @param request The search request.
+     * @param currentPage The current page.
+     * @return The search component.
+     */
+    @NotNull
+    private Search getSearchComponent(@NotNull final SlingHttpServletRequest request, @NotNull final Page currentPage) {
+        String suffix = request.getRequestPathInfo().getSuffix();
+        String relativeContentResource = Optional.ofNullable(suffix)
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/SearchResultServlet.java:180]: I'd rename it `relativeContentResourcePath`

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/SearchResultServlet.java [hunk 2]
```diff
@@ -16,282 +16,352 @@
...
-            LOGGER.error("cannot serialize to JSON",e);
-        }
+    /**
+     * Gets the search component resource from the page. Looks inside experience fragments in the page too.
+     *
+     * @param pageResource The page resource.
+     * @param relativeContentResource The relative path of the search component resource.
+     * @return The search component resource.
+     */
+    private Optional<Resource> getSearchComponentResourceFromPage(@NotNull final Resource pageResource, final String relativeContentResource) {
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/SearchResultServlet.java:201]: I'd rename the argument `relativeContentResourcePath`

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/SearchResultServlet.java [hunk 3]
```diff
@@ -16,282 +16,352 @@
...
-        }
-        return searchContentResource;
+    /**
+     * Gets the search component resource from the page's template. Looks inside experience fragments in the templat too.
+     *
+     * @param currentPage The current page, whose template will be used.
+     * @param relativeCotentResource The relative path of the search component resource.
+     * @return The search component resource.
+     */
+    private Optional<Resource> getSearchComponentResourceFromTemplate(@NotNull final Page currentPage, final String relativeCotentResource) {
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/SearchResultServlet.java:214]: I'd rename the argument `relativeContentResourcePath`

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/SearchResultServlet.java [hunk 4]
```diff
@@ -16,282 +16,352 @@
...
     }
 
+    /**
+     * Gets the search component resource from experience fragments under the resource. Walks down the children tree.
+     *
+     * @param resource The resource where experience fragments with search component would be looked up.
+     * @param relativeContentResource The relative path of the search component resource.
+     * @return The search component resource.
+     */
+    private Optional<Resource> getSearchComponentResourceFromFragments(Resource resource, String relativeContentResource) {
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/SearchResultServlet.java:230]: I'd rename the argument `relativeContentResourcePath`

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/SearchResultServlet.java [hunk 5]
```diff
@@ -16,282 +16,352 @@
...
+    }
+
+    /**
+     * Gets the search component resource from a candidate experience fragment component resource.
+     *
+     * @param candidate The candidate experience fragment component resource.
+     * @param relativeContentResource The relative path of the search component resource.
+     * @return The search component resource.
+     */
+    private Optional<Resource> findFragmentProperties(Resource candidate, String relativeContentResource) {
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/SearchResultServlet.java:247]: I'd rename the argument `relativeContentResourcePath`

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/SearchResultServlet.java [hunk 6]
```diff
@@ -16,282 +16,352 @@
...
+        return Optional.ofNullable(candidate)
+            .map(Resource::getValueMap)
+            .filter(properties -> properties.containsKey(PN_FRAGMENT_VARIATION_PATH))
```
Comment (jckautzmann) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/SearchResultServlet.java:250]: IMHO this line is not needed

**Review by bpauli — APPROVED** [review]

**Pull request merged**